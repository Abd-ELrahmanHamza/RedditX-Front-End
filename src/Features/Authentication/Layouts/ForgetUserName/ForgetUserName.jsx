import { useState, useEffect } from "react";

import { SiReddit } from "react-icons/si";

import ReCAPTCHA from "react-google-recaptcha";

import FormInputPageCom from "Features/Authentication/Components/FormInputPageCom/FormInputPageCom";

import Button from "Features/Authentication/Components/Button/Button";

import {
  AuthContainer,
  ButtonsContainer,
  ErrorParagraph,
  AuthParagraph,
  AuthHeader,
  Forget,
  ReCAPTCHAContainer,
  RedditIcon,
} from "./ForgetUserName.styled";

const USER_EMAIL =
  /[a-zA-Z0-9._-]{3,}@[a-zA-Z0-9._-]{3,}[.]{1}[a-zA-Z0-9._-]{2,}/;

const defaultFormFields = {
  email: "",
};

/**
 * ForgetUserName component that is used in forget userName page
 * @returns {React.Component}  ForgetUserName component that is used in forget userName page
 */

const ForgetUserName = () => {
  /**
   * state to know if the email is valid or not to control what to show to the user
   */
  const [validEmail, setValidEmail] = useState(false);

  /**
   * state to know if the user passed the captcha or not
   */
  const [notRobot, setNotRobot] = useState(false);

  /**
   * state to handel any change the user make in the input fields
   */
  const [formFields, setFormFields] = useState(defaultFormFields);

  /**
   * state to know if the user focused on input field at least one time
   */
  const [initialFocus, setInitialFocus] = useState(true);

  const { email } = formFields;

  /**
   * some configurations for the captcha
   */
  const grecaptchaObject = window.grecaptcha;
  window.recaptchaOptions = {
    useRecaptchaNet: true,
  };

  /**
   * useEffect for email field to check if the email that the user entered is valid or not
   */

  useEffect(() => {
    setValidEmail(USER_EMAIL.test(email));
  }, [email]);

  /**
   * Function to handle the submit of the form of forget userName
   * @param {*} event
   */
  const handleSubmit = async (event) => {
    event.preventDefault();
  };

  /**
   * Function to handle any change on the input field of the forget userName form (check if the email is valid or not)
   * @param {*} event
   */

  const handleChange = (event) => {
    const { name, value } = event.target;

    if (name === "email") {
      setValidEmail(USER_EMAIL.test(email));
    }

    setFormFields({ ...formFields, [name]: value });
  };


  /**
   * Function to handle the captcha and set the state of notRobot to true
   * @param {boolean} value the value of captcha
   */
  const recaptchaHandler = (value) => {
    setNotRobot(true);
  };

  return (
    <AuthContainer>
      <RedditIcon>
        <SiReddit size={60} color={"red"}></SiReddit>
      </RedditIcon>
      <AuthHeader>Recover your username</AuthHeader>
      <AuthParagraph>
        Tell us the email address associated with your Reddit account, and weâ€™ll
        send you an email with your username.
      </AuthParagraph>
      <br></br>
      <form onSubmit={handleSubmit}>
        <FormInputPageCom
          valid={validEmail}
          initialFocus={initialFocus}
          showIcon={true}
          label="EMAIL"
          type="email"
          required
          onChange={handleChange}
          name="email"
          value={email}
          onFocus={() => {
            setInitialFocus(false);
          }}
        />

         {/* Show error message if the email is not valid and the user made a focus on the it's input field */}
        <ErrorParagraph valid={validEmail || initialFocus}>
          Please fix your email to continue
        </ErrorParagraph>
        <br></br>
        <ButtonsContainer>

          {/** Show an enabled button if the email is valid and the user has passed the captcha*/}
          {validEmail && notRobot && (
            <Button
              page={true}
              valid={validEmail}
              type="button"
              onClick={() => {
                console.log("Enabled");
              }}
            >
              EMAIL ME
            </Button>
          )}

          {/** Show an disabled button if the email is not valid or the user has not passed the captcha*/}
          {(!validEmail || !notRobot) && (
            <Button
              page={true}
              disabled
              valid={validEmail && notRobot}
              type="button"
            >
              EMAIL ME
            </Button>
          )}
        </ButtonsContainer>
       
          <ReCAPTCHAContainer validEmail={validEmail}>
            <ReCAPTCHA
              sitekey="6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI"
              onChange={recaptchaHandler}
              grecaptcha={grecaptchaObject}
            />
          </ReCAPTCHAContainer>
       

        <Forget>
          Don't have an email or need assistance logging <br /> in?
          <button onClick={() => {}}>GET HELP</button>{" "}
        </Forget>
        <br></br>
        <br></br>
        <Forget>
          {/* Want to login? */}
          <span onClick={() => {}}>LOG IN</span>{" "}
          <span onClick={() => {}}>{" . "}SIGN UP</span>{" "}
        </Forget>
      </form>
    </AuthContainer>
  );
};

export default ForgetUserName;
