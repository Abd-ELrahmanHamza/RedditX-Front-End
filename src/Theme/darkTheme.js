/**
 * dark theme colors
 *
 * @type {{id:string}}
 */
const darkTheme = {
  // The id of the theme
  id: "dark",
  // The canvas color is the background color of the website
  canvas: "#030303",
  // The colors of the text on the website
  color: {
    // The primary color of the text on the website
    primary: "#d7dadc",
    // The secondary color of the website looks like primary in dark theme
    secondary: "#d7dadc",
    // The color of the muted text on the website
    muted: "#7c7c7c",
    // The color of the links on the website
    link: "#0079d3",
    // The orange color of the website
    orange: "#d93a00",
    // The pink color of the website
    pink: "#fb133a",
    // The color of the text of the main links of search page
    titleText_of_link: "#D7DADC",

    // Some colors for authentication part
    logInButtonColorHovered: "#dc4a14",
    logInBlueHovered: "#3394dc",
    baseButtonColor: "#fff",
    subColor: "#797b7d",
    mainInputColor: "black",
    borderColor: "#86888a",
    blueValid: "#24a0ed",
    mainInputBackGroundColor: "#f6f7f8",
    initialBorderColor: "#e6e6e5",
    initialLabelColor: "#aeadad",
    errorColor: "#ea0027",
    orHeaderColor: "#93908e",
    authParagraphColor: "#1a1a1b",
    signInWithFacebookColor: "#237df3",
  },
  // The colors of the backgrounds in the website
  background: {
    // The primary color of the background on the website
    primary: "#1a1a1b",
    // The color of the muted background on the website
    muted: "#7c7c7c3d",
    // The pink color of the website
    pink: "#fb133a",
    // The light muted color of the website
    lightMuted: "#f6f7f8",
  },
  // Border colors
  borderColor: {
    // Primary border color
    primary: "#818384",
    // Secondary border color
    secondary: "#343536",
    // Muted border color
    muted: "#878a8c80",
    // The button hover color of the search page
    fill: "#2D2D2F",
    // The button active color of the search page
    active: "#272729",
    // button background color hover of the search page cards
    card_hover: "#353537",
    // Comment background color in search page
    comment: "#17232D",
  },
  // The colors of the lines in the website
  lineColor: {
    primary: "#343536",
    // Post Border Line Color Search Page
    PostBorder: "#343536",
  },
  // The colors of the buttons in the website
  button: {
    // The color when hover on buttons on the website
    muted: "#1a1a1b",
  },
};

export default darkTheme;
