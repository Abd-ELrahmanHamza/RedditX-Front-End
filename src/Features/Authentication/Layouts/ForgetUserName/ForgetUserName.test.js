import { screen, render, fireEvent, waitFor } from "@testing-library/react";
import React from "react";
import ForgetUserName from "./ForgetUserName";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";

import darkTheme from "Theme/darkTheme";


const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

describe("ForgetUserName Layout", () => {
  test("email-error shouldn't be found", () => {
    render(
      <BrowserRouter>
        <ThemeProvider theme={darkTheme}>
          <ForgetUserName />
        </ThemeProvider>
      </BrowserRouter>
    );
    expect(screen.queryByTestId("email-error")).not.toBeVisible();
  });

  test("email-error should be found", async () => {
    render(
      <BrowserRouter>
        <ThemeProvider theme={darkTheme}>
          <ForgetUserName />
        </ThemeProvider>
      </BrowserRouter>
    );
    const emailInput = screen.getByTestId("email");

    fireEvent.change(emailInput, {
      target: { value: "M" },
    });

    emailInput.focus();

    const emailError = screen.getByTestId("email-error");

    await waitFor(() => {
      expect(emailError).toBeVisible();
    });
  });
  test("email-error should be not found", async () => {
    render(
      <BrowserRouter>
        <ThemeProvider theme={darkTheme}>
          <ForgetUserName />
        </ThemeProvider>
      </BrowserRouter>
    );
    const emailInput = screen.getByTestId("email");

    fireEvent.change(emailInput, {
      target: { value: "Mgfdgs@gmail.com" },
    });

    const emailError = screen.getByTestId("email-error");

    await waitFor(() => {
      expect(emailError).not.toBeVisible();
    });
  });

  test("email-input should be found", () => {
    render(
      <BrowserRouter>
        <ThemeProvider theme={darkTheme}>
          <ForgetUserName />
        </ThemeProvider>
      </BrowserRouter>
    );
    const emailInput = screen.getByTestId("email");
    expect(emailInput).toBeInTheDocument();
  });

});
