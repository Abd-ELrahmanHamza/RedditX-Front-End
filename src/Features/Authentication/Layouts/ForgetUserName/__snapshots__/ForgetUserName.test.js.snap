// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgetUserName Layout this is a test for ForgetUserName Layout 1`] = `ShallowWrapper {}`;
